rush.md

```diff
@@ -0,0 +1,374 @@
+
+### 2.3 Computing the Ordering from Dag.
+
+So far (in Section 2.2) we learned that the main object in Aleph is the ever-growing Dag composed of units created by different nodes. We ensure (and thus assume this from now on) that all the honest nodes, at every moment in time, hold a reasonably up-to-date version of the Dag. We now proceed to describe an algorithm `OrderData`, which given a Dag `D` (a static object), computes a sequence (for instance `Vec<Data>` in Rust) of Data items (coming from some units in the Dag) which has the following fundamental **MONOTONICITY** property:
+
+ **If `D` is a subset of `D'` then  `OrderData(D)` is a prefix of `OrderData(D')`.**
+
+ Let us pause for a minute to realize how strong the property is, and that it already solves the consensus problem. Let us denote by `D_1`, `D_2`, ..., `D_N` the local Dags of nodes `1, 2, ..., N`, and by `order_1`, `order_2`, ..., `order_N` the results of applying the `OrderData` algorithm to these Dags. Further let `D` be the a dag being the union of all `D_i`'s for `i=1, 2, ..., N`. From the property above it follows that if `order = OrderData(D)` then each `order_i` is a prefix of `order`, and hence consequently `order_i` also agree with each other! Given now such an algorithm `OrderData` we can complete the description of the Aleph protocol: each node simply keeps its local dag `D` up-to-date and recomputes `OrderData(D)` over and over, the ever-growing sequence od Data items obtained from this procedure is this node's output stream.
+
+ **Note:** *The description of `OrderData` provided below is nowhere near to what happens in Rush's code. This description is optimized for ease of understanding the main concept. An efficient implementation of the idea is another pair of shoes. In fact, the name `OrderData` is not even used in the code -- the ordered data instead lands in an output stream that is produced asynchronously.*
+
+
+#### 2.3.1 Description of OrderData.
+
+For a unit `U` in a certain dag let us denote by `below(U)` the set of all units that can be reached from `U` by following 0 or more forward edges from `U`, so this set consists of `U`, parents of `U`, parents of parents of `U`, and so on. Below we describe `OrderData(D)` based on a procedure `Head(r, D)` that given a round `r` and dag `D` outputs an element of type `Option<Unit>` thus either `None` or `Some(U)` where `U` is some unit in `D` of round `r`. The description of  `Head(r, D)` follows later.
+
+```
+def OrderData(D):
+	order = []
+	batch = []
+	for r = 0, 1, 2, ... :
+		if Head(r, D) = Some(U):
+			new_batch = below(U)
+			remove from new_batch the contents of batch[0], batch[1], ..., batch[r-1]
+			sort new_batch using any canonical ordering
+			batch[r] = new_batch
+			let [U_1, U_2, ..., U_k] = new_batch
+			order = order ++ [U_1.data, U_2.data, ..., U_k.data]
+		else:
+			break
+	return order
+```
```

Comment:
Generally I agree that Some() + None is not the standard pseudocode language, but in this case it allows to formulate things succinctly and clearly and I went for it anyway (there is also some explanation of this syntax above the pseudocode). 

Atfer your comment I started changing the code but rolled back quickly, because it would require to change too many things and for worse. One example: how would you formulate the "monotonicity" property without making it long or confusing using the pythonic syntax?

Reply:
I find `if Head(r, D) = Some(U):` confusing... I have never seen a notation where the pattern of a pattern matching is on the right hand side of `=`.

How about:
```
    let head = Head(r, D)
    if head == None:
        break
    let Some(U) = head
    ...
```
?

And I don't claim that all parts should use pythonic syntax. Natural language parts are awesome.

Reply:
Right, I think this makes sense -- I changed it according to your suggestion. Now the pattern matching is hopefully more clear.